File: projects/chess-coach/apps/harness/src/test/kotlin/harness/OrchestratorTest.kt
package harness

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.runBlocking
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class OrchestratorTest {
    @Test
    fun `generateAdvice uses pre-move prompt semantics and returns commentary only`() = runBlocking {
        val stateManager = StateManager()
        val engineBridge = FakeEngineBridge(
            evaluations = mapOf(
                START_FEN to EvalResult(bestMove = "g1f3", cp = 25, isMate = false, mateIn = 0, pv = ""),
                AFTER_E4_FEN to EvalResult(bestMove = "g8f6", cp = 18, isMate = false, mateIn = 0, pv = "")
            ),
            fenTransitions = mapOf(
                START_FEN to mapOf("e4" to AFTER_E4_FEN)
            )
        )
        val llmClient = RecordingLlmClient(
            promptResponse = """
                Commentary: Nice move controlling the center.
                Predicted ELO: 1750
                Verified Classification: Good Move
            """.trimIndent()
        )

        val orchestrator = TestOrchestrator(stateManager, engineBridge, llmClient)

        val result = orchestrator.generateAdvice("e4", "", START_FEN)

        assertEquals("Nice move controlling the center.", result)
        assertEquals(EngineConfig.Llm.SYSTEM_PROMPT, llmClient.lastSystemPrompt)
        assertTrue(llmClient.lastUserPrompt.contains("Type: standard"))
        assertTrue(llmClient.lastUserPrompt.contains("FEN: $START_FEN"))
        assertTrue(llmClient.lastUserPrompt.contains("MoveSAN: e4"))
        assertTrue(llmClient.lastUserPrompt.contains("Side: White"))
        assertTrue(llmClient.lastUserPrompt.contains("BestAlt: g1f3"))
        assertTrue(llmClient.lastUserPrompt.contains("CP: 25->18 (Δ=7)"))
    }

    @Test
    fun `generateExplanation uses actual moveSan and returns commentary only`() = runBlocking {
        val stateManager = StateManager()
        val engineBridge = FakeEngineBridge(
            evaluations = mapOf(
                START_FEN to EvalResult(bestMove = "g1f3", cp = 40, isMate = false, mateIn = 0, pv = ""),
                AFTER_E4_FEN to EvalResult(bestMove = "g8f6", cp = 10, isMate = false, mateIn = 0, pv = "")
            ),
            fenTransitions = emptyMap()
        )
        val llmClient = RecordingLlmClient(
            promptResponse = """
                Commentary: e4 weakens key squares and was less accurate than developing a knight.
                Predicted ELO: 1600
                Verified Classification: Inaccuracy
            """.trimIndent()
        )

        val orchestrator = TestOrchestrator(stateManager, engineBridge, llmClient)

        val result = orchestrator.generateExplanation(
            fenBefore = START_FEN,
            fenAfter = AFTER_E4_FEN,
            moveSan = "e4",
            isBlunder = false
        )

        assertEquals("e4 weakens key squares and was less accurate than developing a knight.", result)
        assertTrue(llmClient.lastUserPrompt.contains("Type: explanation"))
        assertTrue(llmClient.lastUserPrompt.contains("FEN: $START_FEN"))
        assertTrue(llmClient.lastUserPrompt.contains("MoveSAN: e4"))
        assertTrue(llmClient.lastUserPrompt.contains("Side: White"))
        assertTrue(llmClient.lastUserPrompt.contains("BestAlt: g1f3"))
        assertTrue(llmClient.lastUserPrompt.contains("CP: 40->10 (Δ=30)"))
    }

    @Test
    fun `generateAdviceStream emits commentary only`() = runBlocking {
        val stateManager = StateManager()
        val engineBridge = FakeEngineBridge(
            evaluations = mapOf(
                START_FEN to EvalResult(bestMove = "g1f3", cp = 12, isMate = false, mateIn = 0, pv = ""),
                AFTER_E4_FEN to EvalResult(bestMove = "g8f6", cp = 8, isMate = false, mateIn = 0, pv = "")
            ),
            fenTransitions = mapOf(
                START_FEN to mapOf("e4" to AFTER_E4_FEN)
            )
        )
        val llmClient = RecordingLlmClient(
            promptResponse = "",
            streamResponse = listOf(
                "Commentary: Nice move ",
                "grabbing space in the center.\n",
                "Predicted ELO: 1700\n",
                "Verified Classification: Good Move\n"
            )
        )

        val orchestrator = TestOrchestrator(stateManager, engineBridge, llmClient)

        val emitted = mutableListOf<String>()
        orchestrator.generateAdviceStream("e4", "", START_FEN).collect { emitted += it }

        assertEquals(listOf("Nice move grabbing space in the center."), emitted)
    }

    @Test
    fun `generateExplanationStream emits commentary only`() = runBlocking {
        val stateManager = StateManager()
        val engineBridge = FakeEngineBridge(
            evaluations = mapOf(
                START_FEN to EvalResult(bestMove = "g1f3", cp = 80, isMate = false, mateIn = 0, pv = ""),
                AFTER_E4_FEN to EvalResult(bestMove = "g8f6", cp = -40, isMate = false, mateIn = 0, pv = "")
            ),
            fenTransitions = emptyMap()
        )
        val llmClient = RecordingLlmClient(
            promptResponse = "",
            streamResponse = listOf(
                "Commentary: e4 allows too much counterplay.\n",
                "Predicted ELO: 1500\n",
                "Verified Classification: Mistake\n"
            )
        )

        val orchestrator = TestOrchestrator(stateManager, engineBridge, llmClient)

        val emitted = mutableListOf<String>()
        orchestrator.generateExplanationStream(START_FEN, AFTER_E4_FEN, isBlunder = false, moveSan = "e4")
            .collect { emitted += it }

        assertEquals(listOf("e4 allows too much counterplay."), emitted)
    }

    private companion object {
        const val START_FEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
        const val AFTER_E4_FEN = "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1"
    }
}

private class TestOrchestrator(
    stateManager: StateManager,
    engineBridge: FakeEngineBridge,
    llmClient: RecordingLlmClient
) : Orchestrator(
    stateManager = stateManager,
    engineBridge = engineBridge,
    llmClient = llmClient
)

private class FakeEngineBridge(
    private val evaluations: Map<String, EvalResult>,
    private val fenTransitions: Map<String, Map<String, String>>
) : EngineBridge("stockfish") {
    override fun getEvaluation(fen: String, depth: Int): EvalResult {
        return evaluations[fen] ?: error("Missing fake evaluation for FEN: $fen")
    }

    override fun getFenAfterMove(fen: String, move: String): String? {
        return fenTransitions[fen]?.get(move)
    }

    override fun getMaiaMove(fen: String, difficulty: String): String {
        return "e7e5"
    }
}

private class RecordingLlmClient(
    private val promptResponse: String,
    private val streamResponse: List<String> = emptyList()
) : LlmClient("test-model") {
    var lastSystemPrompt: String = ""
    var lastUserPrompt: String = ""

    override suspend fun prompt(
        systemPrompt: String,
        userPrompt: String,
        model: String,
        temperature: Double,
        maxTokens: Int
    ): String {
        lastSystemPrompt = systemPrompt
        lastUserPrompt = userPrompt
        return promptResponse
    }

    override suspend fun promptStream(
        systemPrompt: String,
        userPrompt: String,
        model: String,
        temperature: Double,
        maxTokens: Int
    ): Flow<String> {
        lastSystemPrompt = systemPrompt
        lastUserPrompt = userPrompt
        return flow {
            streamResponse.forEach { emit(it) }
        }
    }
}

File: projects/chess-coach/apps/server/src/Orchestrator.test.ts
import { expect, test } from "bun:test";

import type { EvalResult, IEngineBridge } from "./EngineBridge.ts";
import type { LlmClient } from "./LlmClient.ts";
import { Orchestrator } from "./Orchestrator.ts";
import { StateManager } from "./StateManager.ts";

const START_FEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";
const AFTER_E4_FEN = "rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1";
const AFTER_E4_E5_FEN = "rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2";

class FakeEngineBridge implements IEngineBridge {
  constructor(
    private readonly evals: Record<string, EvalResult>,
    private readonly transitions: Record<string, Record<string, string>>,
  ) {}
  async getEvaluation(fen: string): Promise<EvalResult> {
    const e = this.evals[fen];
    if (!e) throw new Error(`Missing fake evaluation for FEN: ${fen}`);
    return e;
  }
  async getFenAfterMove(fen: string, move: string): Promise<string | null> {
    return this.transitions[fen]?.[move] ?? null;
  }
  async getMaiaMove(): Promise<string> {
    return "e7e5";
  }
}

class RecordingLlmClient implements LlmClient {
  readonly commentaryModel = "test-model";
  lastSystemPrompt = "";
  lastUserPrompt = "";
  constructor(
    private readonly promptResponse: string,
    private readonly streamChunks: string[] = [],
  ) {}
  async prompt(systemPrompt: string, userPrompt: string): Promise<string> {
    this.lastSystemPrompt = systemPrompt;
    this.lastUserPrompt = userPrompt;
    return this.promptResponse;
  }
  async *promptStream(systemPrompt: string, userPrompt: string): AsyncGenerator<string, void, unknown> {
    this.lastSystemPrompt = systemPrompt;
    this.lastUserPrompt = userPrompt;
    for (const c of this.streamChunks) yield c;
  }
}

const tmpState = () => new StateManager(`/tmp/cc-test-${Math.floor(performance.now())}-${process.pid}.fen`);

test("executeTurn returns the Maia move and resulting FEN", async () => {
  const bridge = new FakeEngineBridge(
    { [AFTER_E4_FEN]: { bestMove: "g1f3", cp: 20, isMate: false, mateIn: 0 } },
    { [AFTER_E4_FEN]: { e7e5: AFTER_E4_E5_FEN } },
  );
  const sm = tmpState();
  sm.writeFen(AFTER_E4_FEN);
  const orch = new Orchestrator(sm, bridge, new RecordingLlmClient(""));
  const result = await orch.executeTurn("e4", "intermediate");
  expect(result.move).toBe("e7e5");
  expect(result.fen).toBe(AFTER_E4_E5_FEN);
});

test("executeTurn throws on an illegal Maia move", async () => {
  const bridge = new FakeEngineBridge(
    { [AFTER_E4_FEN]: { bestMove: "g1f3", cp: 20, isMate: false, mateIn: 0 } },
    {}, // no transition => getFenAfterMove returns null => illegal
  );
  const sm = tmpState();
  sm.writeFen(AFTER_E4_FEN);
  const orch = new Orchestrator(sm, bridge, new RecordingLlmClient(""));
  await expect(orch.executeTurn("e4", "intermediate")).rejects.toThrow("illegal move");
});

test("generateAdviceStream emits commentary only with pre-move prompt semantics", async () => {
  const bridge = new FakeEngineBridge(
    {
      [START_FEN]: { bestMove: "g1f3", cp: 25, isMate: false, mateIn: 0 },
      [AFTER_E4_FEN]: { bestMove: "g8f6", cp: 18, isMate: false, mateIn: 0 },
    },
    { [START_FEN]: { e4: AFTER_E4_FEN } },
  );
  const llm = new RecordingLlmClient("", [
    "Commentary: Nice move ",
    "grabbing space in the center.\n",
    "Predicted ELO: 1700\n",
    "Verified Classification: Good Move\n",
  ]);
  const orch = new Orchestrator(tmpState(), bridge, llm);

  const emitted: string[] = [];
  for await (const c of orch.generateAdviceStream("e4", "", START_FEN)) emitted.push(c);

  expect(emitted).toEqual(["Nice move grabbing space in the center."]);
  expect(llm.lastUserPrompt).toContain("Type: standard");
  expect(llm.lastUserPrompt).toContain(`FEN: ${START_FEN}`);
  expect(llm.lastUserPrompt).toContain("MoveSAN: e4");
  expect(llm.lastUserPrompt).toContain("Side: White");
  expect(llm.lastUserPrompt).toContain("BestAlt: g1f3");
  expect(llm.lastUserPrompt).toContain("CP: 25->18 (Δ=7)");
  expect(llm.lastUserPrompt).not.toContain("Instruction:");
});

test("generateExplanationStream emits commentary only with explanation prompt + Instruction", async () => {
  const bridge = new FakeEngineBridge(
    {
      [START_FEN]: { bestMove: "g1f3", cp: 80, isMate: false, mateIn: 0 },
      [AFTER_E4_FEN]: { bestMove: "g8f6", cp: -40, isMate: false, mateIn: 0 },
    },
    {},
  );
  const llm = new RecordingLlmClient("", [
    "Commentary: e4 allows too much counterplay.\n",
    "Predicted ELO: 1500\n",
    "Verified Classification: Mistake\n",
  ]);
  const orch = new Orchestrator(tmpState(), bridge, llm);

  const emitted: string[] = [];
  for await (const c of orch.generateExplanationStream(START_FEN, AFTER_E4_FEN, false, "e4")) {
    emitted.push(c);
  }

  expect(emitted).toEqual(["e4 allows too much counterplay."]);
  expect(llm.lastUserPrompt).toContain("Type: explanation");
  expect(llm.lastUserPrompt).toContain("CP: 80->-40 (Δ=120)");
  expect(llm.lastUserPrompt).toContain("Instruction: Write the commentary in future tense.");
});

test("generateUiPhrases returns the canonical phrase lists", async () => {
  const orch = new Orchestrator(tmpState(), new FakeEngineBridge({}, {}), new RecordingLlmClient(""));
  const phrases = await orch.generateUiPhrases();
  expect(phrases.thinking).toEqual([
    "Hmm...",
    "Let me think...",
    "Interesting position...",
    "Rats...",
    "What to do...",
  ]);
  expect(phrases.bestMove).toEqual(["Great move!", "Excellent!", "I like that.", "Strong play.", "Well done."]);
});
